### PR TITLE
ENH: Add rolling WLS and OLS

### LINKED ISSUE #2302: ENH: Migrate pandas MovingOLS
Part of moving the pandas stats stuff to statsmodels to support.

see also #2745

### LINKED ISSUE #2745: SUMM/ENH: migrate pandas stats models
see also #2302 for MovingOLS

The models in pandas.stats have now been deprecated. https://github.com/pydata/pandas/issues/6077

We should add some things in statsmodels and check that we have roughly the same functionality
- fama_macbeth: there is no equivalent in statsmodels, example and test case is in pandas issue
  https://github.com/pydata/pandas/issues/6077#issuecomment-34592683
  I guess we can initially move it largely the way it is, get it to work, and refactor later if we want a more generic solution. (IIRC, I have seen similar models in other contexts.)
- OLS obsolete, I guess we have already everything covered (and more than pandas)
- Moving/Rolling and Expanding OLS:  I haven't looked in a long time. 
  Ignore and let users use generic rolling methods, or write specific code?  
- Panel data: AFAIR, pandas only has fixed effects, which is just OLS with dummies (e.g. use patsy). 
  We have open issues to concentrate out the fixed effects dummies #2568, see also #2710 (sparse matrices)
  Panel robust covariance matrices need some cleanup, independently of this.
  Econometrics panel data support with two random effects is in PR. #1133
  MixedLM has more general mixed effects structure that also covers now crossed effects.
- VAR: I guess it's covered by VAR in statsmodels, but I never compared.

Not clear: missing value handling, balanced panel structure, there might be some details or options that don't work the same way.

### PR DESCRIPTION
**Pull request by bashtage**
Provide a rolling estimator similar in spirit to the old pandas version

closes #2745
closes #2302

- [X] closes #2745
- [X] closes #2302
- [X] tests added / passed. 
- [X] code/documentation is well formatted.  
- [X] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in master and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/master -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by bashtage** [comment]
Small number of todo:

- [ ] Formula in notebook
- [ ] Complete test coverage of edge cases

File: statsmodels/base/model.py [hunk 1]
```diff
@@ -2130,7 +2127,7 @@ def load(cls, fname):
         Parameters
         ----------
         fname : {str, handle}
```
Comment (jbrockmendel) [inline statsmodels/base/model.py]: i think the braces pattern is supposed to be only for literals

File: statsmodels/regression/tests/test_rolling.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            if isinstance(ci_cols, pd.DataFrame):
+                ci_val = ci_cols.iloc[i - 1]
+                ci_val = np.asarray(ci_val).reshape((-1, 2))
+            else:
+                ci_val = ci_cols[i - 1].T
+            assert_allclose(ci_val, wls_ci)
+
+
+# 'conf_int',
+# load remove_data save
```
Comment (jbrockmendel) [inline statsmodels/regression/tests/test_rolling.py]: i dont understand these comments

File: statsmodels/regression/tests/test_rolling.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+@pytest.mark.parametrize('cov_type', ['nonrobust', 'HC0'])
+@pytest.mark.parametrize('use_t', [None, True, False])
+def test_against_wls_inference(data, use_t, cov_type):
+    y, x, w = data
+    mod = RollingWLS(y, x, window=100, weights=w)
+    res = mod.fit(use_t=use_t, cov_type=cov_type)
+    ci_cols = ci = res.conf_int()
+    test_cols = x.shape[1] > 3
+    # This is a smoke to make sure it works
```
Comment (jbrockmendel) [inline statsmodels/regression/tests/test_rolling.py]: add whitespace before the comment and after cov_params call to clarify that is what the comment applies to

**Comment by jbrockmendel** [comment]
haven't yet looked at the rolling module itself.  it would be easier if you split off the non-rollling parts of the PR, which generally look good

File: statsmodels/regression/tests/test_rolling.py (hunk 2 above)
Comment (bashtage) [reply statsmodels/regression/tests/test_rolling.py]: My notes forgot to remove.

**Comment by bashtage** [comment]
> haven't yet looked at the rolling module itself. it would be easier if you split off the non-rollling parts of the PR, which generally look good

I split it.  This PR is rolling only now.
